File: cpp/data_structures/graph/adjacency_list_dfs.cpp
#include <iostream>
#include <vector>
#include <unordered_set>

class Graph {
public:
    Graph(int vertices);
    void addEdge(int v, int w);
    void DFS(int startVertex);

private:
    int vertices;
    std::vector<std::vector<int>> adjacencyList;

    void DFSUtil(int vertex, std::unordered_set<int>& visited);
};

Graph::Graph(int vertices) : vertices(vertices) {
    adjacencyList.resize(vertices);
}

void Graph::addEdge(int v, int w) {
    adjacencyList[v].push_back(w);
    adjacencyList[w].push_back(v);
}

void Graph::DFSUtil(int vertex, std::unordered_set<int>& visited) {
    visited.insert(vertex);
    std::cout << vertex << " ";

    for (int adjacentVertex : adjacencyList[vertex]) {
        if (visited.find(adjacentVertex) == visited.end()) {
            DFSUtil(adjacentVertex, visited);
        }
    }
}

void Graph::DFS(int startVertex) {
    std::unordered_set<int> visited;

    std::cout << "DFS starting from vertex " << startVertex << ": ";
    DFSUtil(startVertex, visited);

    std::cout << std::endl;
}

int main() {
    // Create a simple graph
    Graph myGraph(6);

    myGraph.addEdge(0, 1);
    myGraph.addEdge(0, 2);
    myGraph.addEdge(1, 3);
    myGraph.addEdge(2, 4);
    myGraph.addEdge(2, 5);

    myGraph.DFS(0);

    return 0;
}

File: cs/data_structures/graph/adjacency_list_dfs.cs
using System;
using System.Collections.Generic;
using System.Linq;

class Graph
{
    private Dictionary<int, List<int>> adjacencyList;

    public Graph()
    {
        adjacencyList = new Dictionary<int, List<int>>();
    }

    public void AddEdge(int vertex, int neighbor)
    {
        if (!adjacencyList.ContainsKey(vertex))
        {
            adjacencyList[vertex] = new List<int>();
        }

        adjacencyList[vertex].Add(neighbor);

        // For undirected graph, add the reverse edge
        if (!adjacencyList.ContainsKey(neighbor))
        {
            adjacencyList[neighbor] = new List<int>();
        }

        adjacencyList[neighbor].Add(vertex);
    }

    public void DFS(int startVertex)
    {
        HashSet<int> visited = new HashSet<int>();
        Console.WriteLine("DFS Traversal:");

        DFSRecursive(startVertex, visited);
    }

    private void DFSRecursive(int currentVertex, HashSet<int> visited)
    {
        visited.Add(currentVertex);
        Console.Write(currentVertex + " ");

        foreach (int neighbor in adjacencyList[currentVertex].Where(n => !visited.Contains(n)))
        {
            DFSRecursive(neighbor, visited);
        }
    }
}

class Program
{
    static void Main()
    {
        Graph graph = new Graph();

        // Adding edges to the graph
        graph.AddEdge(0, 1);
        graph.AddEdge(0, 2);
        graph.AddEdge(1, 3);
        graph.AddEdge(1, 4);
        graph.AddEdge(2, 5);
        graph.AddEdge(2, 6);

        // Perform DFS starting from vertex 0
        graph.DFS(0);
    }
}
